ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,664 @@
+                    " respectively."
+                )
+        for self_tp_or_fp, new_tp_or_fp, name in [(self._tp, tps, "TP"), (self._fp, fps, "FP")]:
+            new_tp_or_fp.keys()
+            try:
+                cls = (self_tp_or_fp.keys() & new_tp_or_fp.keys()).pop()
+            except KeyError:
+                pass
+            else:
+                if self_tp_or_fp[cls][-1].shape[:-1] != new_tp_or_fp[cls].shape[:-1]:
+                    raise ValueError(
+                        f"Tensors in returned {name} from do_matching should not change in shape "
+                        "except possibly in the last dimension which is the dimension of samples. Given "
+                        f"{self_tp_or_fp[cls][-1].shape} and {new_tp_or_fp[cls].shape}"
+                    )
+
+    def _classification_prepare_output(
+        self, y_pred: torch.Tensor, y: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Prepares and returns scores and P tensor. Input and output shapes of the method is as follows.
+
+        ========== =========== ============
+        ``y_pred``
+        -----------------------------------
+        Data type  Input shape Output shape
+        ========== =========== ============
+        Binary     (N, ...)    (1, N * ...)
+        Multilabel (N, C, ...) (C, N * ...)
+        Multiclass (N, C, ...) (C, N * ...)
+        ========== =========== ============
+
+        ========== =========== ============
+        ``y``
+        -----------------------------------
+        Data type  Input shape Output shape
+        ========== =========== ============
+        Binary     (N, ...)    (1, N * ...)
+        Multilabel (N, C, ...) (C, N * ...)
+        Multiclass (N, ...)    (N * ...)
+        ========== =========== ============
+        """
+
+        if self._type == "multilabel":
+            num_classes = y_pred.size(1)
+            scores = torch.transpose(y_pred, 1, 0).reshape(num_classes, -1)
+            P = torch.transpose(y, 1, 0).reshape(num_classes, -1)
+        elif self._type == "binary":
+            P = y.view(1, -1)
+            scores = y_pred.view(1, -1)
+        else:  # Multiclass
+            num_classes = y_pred.size(1)
+            if y.max() + 1 > num_classes:
+                raise ValueError(
+                    f"y_pred contains fewer classes than y. Number of classes in prediction is {num_classes}"
+                    f" and an element in y has invalid class = {y.max().item() + 1}."
+                )
+            P = y.view(-1)
+            scores = torch.transpose(y_pred, 1, 0).reshape(num_classes, -1)
+
+        return scores, P
+
+    def do_matching(
+        self, pred: Any, target: Any
+    ) -> Tuple[Dict[int, torch.Tensor], Dict[int, torch.Tensor], Dict[int, int], Dict[int, torch.Tensor]]:
+        r"""
+        Matching logic holder of the metric for detection tasks.
```

Comment:
I'm not sure why we want to have non-implemented public method in this class. Let's remove it here and set it as a private method in ObjectDetectionMAP

Reply:
I set this method public with the assumption that there are many mAP out there depending on the user needs and requirements so we just expose the method and the user implement his/her desired matching logic. But when I thought twice, it seems that this isn't the case. There aren't "many" styles of mAP so we the developers could bear the responsibility to implement few other mAP metrics. Ok, I make it private.